File: docs/src/test/java/sample/javadsl/ProducerExample.java
/*
 * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
 * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
 */
package sample.javadsl;

import akka.Done;
import akka.actor.ActorSystem;
import akka.kafka.*;
import akka.kafka.javadsl.Producer;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.javadsl.Sink;
import akka.stream.javadsl.Source;

import java.util.Arrays;
import java.util.Map;
import java.util.concurrent.CompletionStage;

import com.typesafe.config.Config;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import org.apache.kafka.clients.producer.RecordMetadata;
import org.apache.kafka.common.serialization.StringSerializer;

abstract class ProducerExample {
  protected final ActorSystem system = ActorSystem.create("example");

  protected final Materializer materializer = ActorMaterializer.create(system);

  // #producer
  // #settings
  final Config config = system.settings().config();
  final ProducerSettings<String, String> producerSettings =
      ProducerSettings
          .create(config, new StringSerializer(), new StringSerializer())
          .withBootstrapServers("localhost:9092");
  // #settings
  final KafkaProducer<String, String> kafkaProducer =
      producerSettings.createKafkaProducer();
  // #producer

  protected void terminateWhenDone(CompletionStage<Done> result) {
    result
      .exceptionally(e -> {
        system.log().error(e, e.getMessage());
        return Done.getInstance();
      })
      .thenAccept(d -> system.terminate());
  }
}

class PlainSinkExample extends ProducerExample {
  public static void main(String[] args) {
    new PlainSinkExample().demo();
  }

  public void demo() {
    // #plainSink
    CompletionStage<Done> done =
      Source.range(1, 100)
        .map(number -> number.toString())
        .map(value -> new ProducerRecord<String, String>("topic1", value))
        .runWith(Producer.plainSink(producerSettings), materializer);
    // #plainSink

    terminateWhenDone(done);
  }
}

class PlainSinkWithProducerExample extends ProducerExample {
    public static void main(String[] args) {
        new PlainSinkExample().demo();
    }

    public void demo() {
        // #plainSinkWithProducer
        CompletionStage<Done> done =
            Source.range(1, 100)
                .map(number -> number.toString())
                .map(value -> new ProducerRecord<String, String>("topic1", value))
                .runWith(Producer.plainSink(producerSettings, kafkaProducer), materializer);
        // #plainSinkWithProducer

        terminateWhenDone(done);
    }
}

class ObserveMetricsExample extends ProducerExample {
    public static void main(String[] args) {
        new PlainSinkExample().demo();
    }

    public void demo() {
        // #producerMetrics
        Map<org.apache.kafka.common.MetricName, ? extends org.apache.kafka.common.Metric> metrics =
                kafkaProducer.metrics();// observe metrics
// #producerMetrics
    }
}

class ProducerFlowExample extends ProducerExample {
  public static void main(String[] args) {
    new ProducerFlowExample().demo();
  }

    <KeyType, ValueType, PassThroughType> ProducerMessage.Message<KeyType, ValueType, PassThroughType> createMessage(KeyType key, ValueType value, PassThroughType passThrough) {
        return
                // #singleMessage
                new ProducerMessage.Message<KeyType, ValueType, PassThroughType>(
                        new ProducerRecord<>("topicName", key, value),
                        passThrough
                );
        // #singleMessage

    }

    <KeyType, ValueType, PassThroughType> ProducerMessage.MultiMessage<KeyType, ValueType, PassThroughType> createMultiMessage(KeyType key, ValueType value, PassThroughType passThrough) {
        return
                // #multiMessage
                new ProducerMessage.MultiMessage<KeyType, ValueType, PassThroughType>(
                        Arrays.asList(
                                new ProducerRecord<>("topicName", key, value),
                                new ProducerRecord<>("anotherTopic", key, value)
                        ),
                        passThrough
                );
        // #multiMessage

    }

    <KeyType, ValueType, PassThroughType> ProducerMessage.PassThroughMessage<KeyType, ValueType, PassThroughType> createPassThroughMessage(KeyType key, ValueType value, PassThroughType passThrough) {

        ProducerMessage.PassThroughMessage<KeyType, ValueType, PassThroughType> ptm =
                // #passThroughMessage
        new ProducerMessage.PassThroughMessage<>(
                passThrough
        );
        // #passThroughMessage
        return ptm;
    }

  public void demo() {
    // #flow
      CompletionStage<Done> done =
              Source.range(1, 100)
                      .map(number -> {
                          int partition = 0;
                          String value = String.valueOf(number);
                          ProducerMessage.Envelope<String, String, Integer> msg =
                                  new ProducerMessage.Message<String, String, Integer>(
                                          new ProducerRecord<>("topic1", partition, "key", value),
                                          number
                                  );
                          return msg;
                      })

                      .via(Producer.flexiFlow(producerSettings))

                      .map(result -> {
                          if (result instanceof ProducerMessage.Result) {
                              ProducerMessage.Result<String, String, Integer> res = (ProducerMessage.Result<String, String, Integer>) result;
                              ProducerRecord<String, String> record = res.message().record();
                              RecordMetadata meta = res.metadata();
                              return meta.topic() + "/" + meta.partition() + " " + res.offset() + ": " + record.value();
                          } else if (result instanceof ProducerMessage.MultiResult) {
                              ProducerMessage.MultiResult<String, String, Integer> res = (ProducerMessage.MultiResult<String, String, Integer>) result;
                              return res.getParts() .stream().map( part -> {
                                  RecordMetadata meta = part.metadata();
                                  return meta.topic() + "/" + meta.partition() + " " + part.metadata().offset() + ": " + part.record().value();
                              }).reduce((acc, s) -> acc + ", " + s);
                          } else {
                              return "passed through";
                          }
                      })
                      .runWith(Sink.foreach(System.out::println), materializer);
      // #flow

    terminateWhenDone(done);
  }
}

File: docs/src/test/scala/sample/scaladsl/ProducerExample.scala
/*
 * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
 * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
 */
package sample.scaladsl

import java.util

import akka.actor.ActorSystem
import akka.kafka.ProducerMessage
import akka.kafka.ProducerSettings
import akka.kafka.scaladsl.Producer
import akka.stream.scaladsl.Source
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.StringSerializer
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.Sink

import scala.concurrent.Future
import akka.Done
import akka.kafka.ProducerMessage.MultiResultPart

import scala.util.{Failure, Success}

trait ProducerExample {
  val system = ActorSystem("example")

  // #producer
  // #settings
  val config = system.settings.config
  val producerSettings =
    ProducerSettings(config, new StringSerializer, new StringSerializer)
      .withBootstrapServers("localhost:9092")
  // #settings
  val kafkaProducer = producerSettings.createKafkaProducer()
  // #producer

  implicit val ec = system.dispatcher
  implicit val materializer = ActorMaterializer.create(system)

  def terminateWhenDone(result: Future[Done]): Unit = {
    result.onComplete {
      case Failure(e) =>
        system.log.error(e, e.getMessage)
        system.terminate()
      case Success(_) => system.terminate()
    }
  }
}

object PlainSinkExample extends ProducerExample {
  def main(args: Array[String]): Unit = {
    // #plainSink
    val done: Future[Done] =
      Source(1 to 100)
        .map(_.toString)
        .map(value => new ProducerRecord[String, String]("topic1", value))
        .runWith(Producer.plainSink(producerSettings))
    // #plainSink

    terminateWhenDone(done)
  }
}

object PlainSinkWithProducerExample extends ProducerExample {
  def main(args: Array[String]): Unit = {
    // #plainSinkWithProducer
    val done = Source(1 to 100)
      .map(_.toString)
      .map(value => new ProducerRecord[String, String]("topic1", value))
      .runWith(Producer.plainSink(producerSettings, kafkaProducer))
    // #plainSinkWithProducer

    terminateWhenDone(done)
  }
}

object ObserveMetricsExample extends ProducerExample {
  def main(args: Array[String]): Unit = {
    // format:off
    // #producerMetrics
    val metrics: util.Map[org.apache.kafka.common.MetricName, _ <: org.apache.kafka.common.Metric] =
      kafkaProducer.metrics() // observe metrics
    // #producerMetrics
    // format:on
    metrics.clear()
  }
}

object ProducerFlowExample extends ProducerExample {

  def createMessage[KeyType, ValueType, PassThroughType](key: KeyType, value: ValueType, passThrough: PassThroughType) =
    {
      // #singleMessage
      new ProducerMessage.Message[KeyType, ValueType, PassThroughType](
        new ProducerRecord("topicName", key, value),
        passThrough
      )
      // #singleMessage
    }

  def createMultiMessage[KeyType, ValueType, PassThroughType](key: KeyType, value: ValueType, passThrough: PassThroughType) =
    {
      import scala.collection.immutable
      // #multiMessage
      new ProducerMessage.MultiMessage[KeyType, ValueType, PassThroughType](
        immutable.Seq(
          new ProducerRecord("topicName", key, value),
          new ProducerRecord("anotherTopic", key, value)
        ),
        passThrough
      )
      // #multiMessage
    }

  def createPassThroughMessage[KeyType, ValueType, PassThroughType](key: KeyType, value: ValueType, passThrough: PassThroughType) = {
    // format:off
    // #passThroughMessage
    new ProducerMessage.PassThroughMessage(
      passThrough
    )
    // #passThroughMessage
    // format:on
  }

  def main(args: Array[String]): Unit = {
    // format:off
    // #flow
    val done = Source(1 to 100)
      .map { number =>
        val partition = 0
        val value = number.toString
        ProducerMessage.Message(
          new ProducerRecord("topic1", partition, "key", value),
          number
        )
      }

      .via(Producer.flexiFlow(producerSettings))

      .map {
        case ProducerMessage.Result(metadata, message) =>
          val record = message.record
          s"${metadata.topic}/${metadata.partition} ${metadata.offset}: ${record.value}"

        case ProducerMessage.MultiResult(parts, passThrough) =>
          parts.map {
            case MultiResultPart(metadata, record) =>
              s"${metadata.topic}/${metadata.partition} ${metadata.offset}: ${record.value}"
          }.mkString(", ")

        case ProducerMessage.PassThroughResult(passThrough) =>
          s"passed through"
      }
      .runWith(Sink.foreach(println(_)))
    // #flow
    // format:on

    terminateWhenDone(done)
  }
}
